lua/nvim-tree/explorer/sorters.lua

```diff
@@ -23,6 +23,31 @@ local function tbl_slice(t, first, last)
   return slice
 end
 
+---Evaluate `sort.folders_first` and `sort.files_first`
+---@param a table
+---@param b table
+---@param third_option_callback function|nil function should return `boolean`. Evaluated only if necessary
+---@return boolean|nil
+local function folders_or_files_first(a, b, third_option_callback)
+  local x = nil
+
+  if a.nodes and not b.nodes then
+    x = true
+  elseif not a.nodes and b.nodes then
+    x = false
+  elseif a.nodes and b.nodes and third_option_callback ~= nil then
+    x = third_option_callback()
+  end
+
+  if x ~= nil then
+    if M.config.sort.files_first then
```

Comment:
Could we flip this around? Test this config at the start of the function?

That would remove the need for the mystery `x`

Reply:
Yeah, I stored in `x` what should be returned if `sort.folders_first` is enabled and reverse it with `sort.files_first`, but now that I think about it maybe something like this would be more readable and intuitive.
```lua
local a_and_not_b, not_a_and_b

if M.config.sort.files_first then
  a_and_not_b = false
  not_a_and_b = true
elseif M.config.sort.folders_first then
  a_and_not_b = true
  not_a_and_b = false
else
  return
end

if a.nodes and not b.nodes then
  return a_and_not_b
elseif not a.nodes and b.nodes then
  return not_a_and_b
elseif a.nodes and b.nodes and third_option_callback ~= nil then
  return third_option_callback()
end
```

Reply:
Much more readable. Could we go simpler?

```lua
if a.nodes and not b.nodes then
  return M.config.sort.folders_first
```

Reply:
Ah of course. How about an early exit?

```lua
if not M.config.sort.files_first and not M.config.sort.folders_first then
  return
end

Reply:
This should do it:

```lua
---Evaluate `sort.folders_first` and `sort.files_first`, files takes precedence.
---@param a table node
---@param b table node
---@param fallback function|nil function should return `boolean`. Evaluated when both are folders.
---@return boolean|nil
local function folders_or_files_first(a, b, fallback)
  if not M.config.sort.folders_first and not M.config.sort.files_first then
    return
  end

  if not a.nodes and b.nodes then
    -- file <> folder
    return M.config.sort.files_first
  elseif a.nodes and not b.nodes then
    -- folder <> file
    return not M.config.sort.files_first
  elseif fallback and a.nodes and b.nodes then
    -- folder <> folder
    return fallback()
  end
end
```

Yes, we could be tricky with ternaries but readability is more important.

Reply:
Oh, I didn't consider to completely ignore `sort.folders_first` but it works so that's fine.